applications/dirgraph_visualization/dirgraph_utils.py

```diff
@@ -0,0 +1,647 @@
+            upstream = [x['params']['upstream_block'] for x in valves],
+            downstream = [x['params']['downstream_block'] for x in valves]
+        ))
+        create_valve_blocks(d, df_vessels, df_valves, name_type)
+
+    if 'chambers' not in d:
+        df_chambers= pd.DataFrame(columns=['name', 'type'])
+
+    else:
+        chambers = d['chambers']
+        df_chambers = pd.DataFrame(dict(
+            name = [x['name'] for x in chambers],
+            type=[x['type'] for x in chambers],
+        ))
+
+        create_chamber_blocks(d, df_chambers, df_junctions_expanded, df_valves)
+
+    create_outlet_bc_blocks(d, df_valves, name_type)
+    create_inlet_bc_blocks(d, df_valves, name_type)
+    create_vessel_blocks(d, df_vessels, df_junctions_expanded, df_valves, name_type)
+    return d
+
+
+
+def create_valve_blocks(d, df_vessels, df_valves, name_type):
+    valve_blocks = {}  # {block_name: block_object}
+
+    if name_type == 'id':
+        vessel_id_map = get_vessel_name_to_vessel_id_map(d)
+
+    # vessel_list = get_vessel_list(d, 'name')
+    bc_name_to_index_map = dirgraph_steady_bc.get_bc_name_to_index_map(d)
+
+    def process_row(row):
+        valve_name = row['name']
+        ups_block = row['upstream']
+        down_block = row['downstream']
+        connecting_block_list = []
+        flow_directions = []
+
+        # Upstream Block Processing
+        if ups_block in bc_name_to_index_map:
+            bc_block_name = "BC" + valve_name + "_inlet"
+            connecting_block_list.append(bc_block_name)
+            flow_directions.append(-1)
+        elif ups_block in df_vessels['name'].values:
+            vessel_id_name = "V" + str(vessel_id_map[ups_block])
+            connecting_block_list.append(vessel_id_name)
+            flow_directions.append(-1)
+        else:
+            connecting_block_list.append(ups_block)
+            flow_directions.append(-1)
+
+        # Downstream Block Processing
+        if down_block in bc_name_to_index_map:
+            bc_block_name = "BC" + valve_name + "_outlet"
+            connecting_block_list.append(bc_block_name)
+            flow_directions.append(+1)
+        elif down_block in df_vessels['name'].values:
+            vessel_id_name = "V" + str(vessel_id_map[down_block])
+            connecting_block_list.append(vessel_id_name)
+            flow_directions.append(+1)
+        else:
+            connecting_block_list.append(down_block)
+            flow_directions.append(+1)
+
+        if (+1 in flow_directions) and (-1 in flow_directions):
+            valve_blocks[valve_name] = ntwku.LPNBlock(
+                connecting_block_list=connecting_block_list,
+                name=valve_name,
+                flow_directions=flow_directions
+            )
+
+    df_valves.apply(process_row, axis=1)
+    d["blocks"].update(valve_blocks)
+
+
+
+def create_chamber_blocks(d, df_chambers, df_junctions_expanded, df_valves):
+    """
```

Comment:
Some functions have these nice comments and some don't. Can you add them to all functions if possible?